### ISSUE TITLE
Have a text template filter in TextFields

### ISSUE THREAD

**Issue by xster** [flutter/flutter#9511]
If text input needs to fit a pattern like phone numbers or credit card, it would be nice to have an API to automatically fill in the parentheses, dashes and spaces as the user types in digits.

**Comment by xster** [comment]
Thinking add to EditableText a this.formatter of type abstract TextInputFormatter. 

It has abstract method TextEditValue formatEditUpdate(TextEditValue oldValue, TextEditValue newValue). 

We provide 2 batteries-included implementations WhitelistingTextInputFormatter and BlacklistingTextInputFormatter. 

Users can roll their own by subclassing. Check in updateEditingValue and call _updateRemoteEditingValueIfNeeded. 

Solves also #9515 with the same API. 

FYI @abarth

**Comment by xster** [comment]
M2 since it blocks multiple M2 issues

**Comment by xster** [comment]
There's some indirections but I think EditableText will handle copy pasting properly too if the formatters are called after _handleSelectionOverlayChanged. 

Though I wonder if calling _updateRemoteEditingValueIfNeeded is really needed.

**Comment by abarth** [comment]
You need to call _updateRemoteEditingValueIfNeeded otherwise the IME will get out-of-sync with the value the user sees.  That will mess up things like selections.

**Comment by xster** [comment]
Cool thanks. Should we call it today on _handleSelectionOverlayChanged? I suspect it's not called now after copy pasting.

**Comment by eseidelGoogle** [comment]
Unless there is a customer blocked on this, I don't think this needs to block M2?

**Comment by slightfoot** [comment]
If you found this ticket wanting this feature you can now specify `inputFormatters` as an argument to `TextFormField` and `TextField`, as well as `EditableText`.

The https://flutter.io/text-input/ docs should be updated to include a bit on this and an example.
